### ISSUE TITLE
[bug?] Form still dirty after FormGroup.reset(values) 

### ISSUE THREAD

**Issue by BrainCrumbz** [kekeh/mydatepicker#295]
Current version: 2.0.0

### Steps

1. Create a reactive form, say named `this.myForm`
1. Add some input control and a `<my-date-picker>` instance to form with their own `formControlName`s
1. After some timeout, invoke `this.myForm.reset(myValues)` passing a correctly shaped object

### Expected

Form input controls should be filled with passed data, and form `dirty` should be false.

### Actual

Form input controls are effectively filled with passed data, but form `dirty` is true. Notice that if one excludes all `<my-date-picker>` controls from the form, the final `dirty` value is correctly false.

---

Sample [plunker here](http://plnkr.co/edit/omXY5e?p=preview), checking *src/my-form.component.ts*

**Comment by BrainCrumbz** [comment]
Note: we also found a previous issue, #182 , but that has a workaround on resetting the single datepicker control. The need here is the more common one of resetting a whole form.

Thanks!

**Comment by kekeh** [comment]
As you probably know in Angular2 there is **ControlValueAccessor** interface: https://angular.io/docs/ts/latest/api/forms/index/ControlValueAccessor-interface.html

This interface contain method **writeValue()**. This method is called in both cases when you call **this.myForm.setValue()** or **this.myForm.reset()**. mydatepicker has implemented this interface.

Problem is that if I update model (you previous issue: https://github.com/kekeh/mydatepicker/issues/291) in this method by calling onChangeCb(myDateModel) it sets dirty to true. It does it bot the **this.myForm.setValue()** or the **this.myForm.reset()** calls.

I think the **writeValue()** method is intented to update only internal model not external. I remove **onChangeCb()** call now from the **writeValue()** method. It affects issue 291 so that mydatepicker works as before the fix 291 but in case of invalid date the value of model in null.

If you have better ideas how to do this let me know.

I did this change in version 2.0.1.

**Comment by BrainCrumbz** [comment]
Could you please reopen this or the other issue, to keep discussion going & visibile?

I don't get what is the strict relationship between the two, in the sense: whether calling `writeValue()` should fire a change callback or not, that is a matter we can discuss about, and I'd like to.

But no matter what's the outcome, when component fires an event, it should adhere to its signature and above all stay consistent. In this case, it means component should condition its input and convert it to the needed internal format, so that it can then later be output with the right shape/type.

We can have a look at some of Material Design custom components implementing `ControlValueAccessor` for some inspiration or guidelines. If e.g. we look at [checkbox](https://github.com/angular/material2/blob/29e503a0f15c7915981be920e7b011b6a4e554e8/src/lib/checkbox/checkbox.ts) we see the following "flow":

[writeValue](https://github.com/angular/material2/blob/29e503a0f15c7915981be920e7b011b6a4e554e8/src/lib/checkbox/checkbox.ts#L279):

```
  writeValue(value: any) {
    this.checked = !!value;
}
```

[set checked](https://github.com/angular/material2/blob/29e503a0f15c7915981be920e7b011b6a4e554e8/src/lib/checkbox/checkbox.ts#L218):
```
  set checked(checked: boolean) {
    if (checked != this.checked) {
      this._checked = checked;
      this._changeDetectorRef.markForCheck();
    }
}
```

This one will perform only if value is actually different. And then invoke `_changeDetectorRef.markForCheck()` ... might be helpful to figure out what that does. It is not directly invoking any other "changed" callback, like `emitChangeEvent` or something similar.

**Comment by kekeh** [comment]
Updating value of model in writeValue function only if the date parameter is not same as existing date could work.

Then when you want to reset you have to give selected date as a parameter of reset function. It keeps dirty false because model do not change.

**Comment by BrainCrumbz** [comment]
Sorry, not sure about your last sentence. [The plunker](http://plnkr.co/edit/omXY5e?p=preview) has been updated to stay stuck at v2.0.0 and to try different test cases:

1. when form HTML does **not** contain datepicker, just plain input, both `setValue({...})` and `reset({...})` emit passed value and leave dirty **false**. Of course, emitted value has always the same shape.

1. when a form HTML **does** contain datepicker together with plain input, invoking both functions will still emit passed value, but will make dirty **true**, which is wrong. Also, sticking with v2.0.0, the emitted shape is the right one.

**Comment by kekeh** [comment]
I have already removed model update in writeValue() method. It means there is not onChangeCb() function call any more.

I don't see any reason to update the value of model in the writeValue() function because setValue() or reset() functions do it automatically.

You can try version 2.0.1 now. I close this issue.

**Comment by BrainCrumbz** [comment]
Yep, that is clea (BTW was mentioned in previous comment too). What happens now, with 2.0.1, to emitted values? Do they have always the same shape or what?

**Comment by BrainCrumbz** [comment]
Just checked with plunker updated to 2.0.1: that other issue is back again, which is what I was trying to figure out in longer comment. 

It's ok that component does not invoke any change callback at that point, but then why the emitted value has a different shape compared to when user clicks on a date. THAT is the inconsistency.

**Comment by kekeh** [comment]
When you call **setValue** you can also set other properties from the **IMyDateModel**. For example set **date** and **jsdate** properties. The **date** property is mandatory and others are optional.
```javascript
setDate(): void {
    // set today
    let date: Date = new Date();
    this.myForm.setValue({myDate: 
        {
            date: 
                 {
                     year: date.getFullYear(), 
                     month: date.getMonth() + 1, 
                     day: date.getDate()
                 }, 
            jsdate: date
        }
    });
}
```
If I fix this problem to date picker then we have again reset() dirty problem.

**Comment by BrainCrumbz** [comment]
Thanks for quick feedback. I see. The point is that, while those are optional fields on the input side, they are "mandatory" (so to say) fields on the output side. 

I don't exactly kwno why it is useful to pass, as an input, both `date` and `jsDate` fields, but I'm sure there's a logic and a rationale behind that. Ideally, when one passes that input, whether it has one or the other field, it should be transformed to the full, correct, internal representation (the `IMyDateModel` one). 

Could you please point to the bit of source code that does "receive" the external input values? And the one responsible for "emitting" output changes?

**Comment by kekeh** [comment]
When you call **setValue** or **reset** functions angular call this: https://github.com/kekeh/mydatepicker/blob/master/src/my-date-picker/my-date-picker.component.ts#L328

it is part of ControlValueAccessor interface. The writeValue function of date picker handles **{date: IMyDate}**, **null** or **empty string**.

Here is registered https://github.com/kekeh/mydatepicker/blob/master/src/my-date-picker/my-date-picker.component.ts#L354 callback which can be used to update model. So calling **this.onChangeCb(IMyDateModel)** updates the model.

Here is registered https://github.com/kekeh/mydatepicker/blob/master/src/my-date-picker/my-date-picker.component.ts#L358 callback which can be used to set component touched value to true. So calling **this.onTouchedCb()** set component to touched.

What I did in fix #291 I added:
```javascript
this.onChangeCb(this.getDateModel(this.selectedDate));
```
line after 338 line. It fixed issue #291 but it also caused this issue.

**Comment by BrainCrumbz** [comment]
Yes, some of those changes were clear by looking at commit history and rollback commits. Do you know where does angular "take" values from components in order to emit those as valueChanges? Because, depending on something still not clear, your component either emits a type of data, or another one. I'd like to figure out why this is happening.
